DOC-734 Add --secrets CLI flag (#101)
redpanda-data/rp-connect-docs: modules/configuration/pages/secrets.adoc

= Secrets
description: Methods for adding secrets to your Redpanda configuration without exposing them.

This topic outlines how to add secrets to a Redpanda Connect configuration without exposing them.

> == Store secrets in environment variables

Author: @Jeffail Could you check changes to this section are accurate?

Reviewer: LGTM

A common way to securely pass secrets to a service is to use environment variables. Redpanda Connect allows you to inject the values of environment variables into a configuration with the interpolation syntax `+${SECRET_ID}+`. For example:

[source,yml]
----
thing:
  super_secret: "${SECRET_ID}"
----

[NOTE]
.Use quotes
====
It is valid to have `+super_secret: ${SECRET}+` above (without the quotes), but if `SECRET` is unset then the configuration becomes structurally different. Therefore, it's always best to wrap environment variable interpolations with quotes so that when the variable is unset you still have a valid configuration (with an empty string).
====

For more information about this syntax, see the xref:configuration:interpolation.adoc[interpolation field page].

> == Look up secrets on a remote system at runtime

Reviewer: LGTM

From version 4.39.0, you can use the `rpk connect` CLI flag `--secrets` to:

- Look up secrets values on a remote system at runtime

> - Set an environment variable value in your configuration

Reviewer: This isn't correct, an environment variable is never set. When a secrets management solution is specified we will use it to directly resolve interpolations without changing anything about the environment.

Author: Updated

For example, you could run the following command to retrieve the value for `"$\{SECRET}"`, when the secret is stored on a Redis server.

```bash

> rpk connect run ./config.yaml --secrets env:redis://secret:redis@example.com

Reviewer 2: ```suggestion
rpk connect run ./config.yaml --secrets redis://secret:redis@example.com
```

Author: Updated

> rpk connect run ./config.yaml --secrets env:redis://secret:redis@example.com

Reviewer: ```suggestion
rpk connect run ./config.yaml --secrets redis://secret:redis@example.com
```

There's no `env:` prefix here, the `env:` secrets URN is its own type, and adds an explicit mechanism for obtaining secrets from environment variables.

Author: Updated.

```

> The `env:` argument tries to load the secret value from the specified URN using the format `scheme://prefix`. The prefix includes the environment variable name, and secret location, which in this case is the Redis server address.

Reviewer 2: The `env:` loads environment variables. `redis:` loads from Redis. The argument has a form `redis://secret:redis@example.com` and we do not support prefixes for Redis.

Author: Updated

Author: @tomasz-sadura and @Jeffail. Is it possible to provide an example command for one of the providers that uses a prefix?

Reviewer 2: for AWS:
`--secrets aws://eu-west-1/redpanda/`
for GCP:
`--secrets gcp://project-id/redpanda-`

Author: Updated table with additional examples.

You can specify multiple URNs separated by commas, which are tried in turn until a secrets value is successfully returned.

> NOTE: If you use the argument `none:` with the `--secrets` flag, the secrets value is still returned, but Redpanda Connect does not set an environment variable.

Reviewer 2: `none:` (as an only one argument to `--secrets`) means that variables will not get resolved at all.

Author: Removed description of `none` argument as doesn't seem relevant.

> NOTE: If you use the argument `none:` with the `--secrets` flag, the secrets value is still returned, but Redpanda Connect does not set an environment variable.

Reviewer: This is incorrect, there is a specific `none:` secrets URN you may specify, and when doing so it explicitly instructs RPCN _not_ to resolve secrets. This is a bit of an odd feature and it's due to the fact that `env:` is quietly there as a default secrets option (otherwise we'd break backwards compatibility). Therefore the only way to disable all forms of interpolation would be to run `rpk connect run --secrets "none:"`.

Author: Removed.

=== Supported remote systems

You can retrieve secrets from all of the following remote systems.

|===
| Remote system | URN format

| AWS
| `aws://region/prefix`

| AZURE
| `az://vault-uri/prefix`. The `vault-uri` value should not contain a `https://` prefix.

| GCP
| `gcp://projectID/prefix`

| Redis

> | `redis://prefix`

Reviewer 2: See above. We do not support prefix for Redis. The URN should be of the form: `redis://secret:redis@example.com`

Author: Updated.

[NOTE]
.Use quotes
====

> It is valid to have `+super_secret: ${SECRET}+` above (without the quotes), but if `SECRET` is unset then the configuration becomes structurally different. Therefore, it's always best to wrap environment variable interpolations with quotes so that when the variable is unset you still have a valid configuration (with an empty string).

Reviewer 3: ```suggestion
It is valid to have `+super_secret: ${SECRET}+` above (without the quotes), but if `SECRET` is unset then the configuration becomes structurally different. Therefore, it's always best to wrap environment variable interpolations within quotes so that when the variable is unset you still have a valid configuration (with an empty string).
```

Author: Updated as requested.

====

For more information about this syntax, see the xref:configuration:interpolation.adoc[interpolation field page].

== Look up secrets on a remote system at runtime

> From version 4.39.0, you can use the `rpk connect` CLI flag `--secrets` to look up secrets values on a remote system at runtime. For example, in your secrets management solution. This means that Redpanda Connect resolves interpolations in your configuration at runtime without setting environment variables.

Reviewer 3: ```suggestion
Starting with version 4.39.0, you can use the `rpk connect` CLI flag `--secrets` to look up secrets values on a remote system at runtime(for example, in your secrets management solution). This means that Redpanda Connect resolves interpolations in your configuration at runtime without setting environment variables.
```

Author: Updated as requested.

For example, you could run the following command to retrieve the value for `"$\{SECRET}"`, when the secret is stored on a Redis server.

```bash

rpk connect run ./config.yaml --secrets redis://secret:redis@example.com

```

> The command tries to load the secret value from the specified URN using the format `scheme://secret:server_address`.

Reviewer 3: URN? Should it be URL? I don't see URN spelled out anywhere here. What is it? Noticed there is no glossary in RPCN docs...maybe time to think about one?

You can specify multiple URNs separated by commas, which are tried in turn until a secrets value is successfully returned.

=== Supported remote systems

You can retrieve secrets from all of the following remote systems.

|===
| Remote system | URN format

| AWS
| `aws://region/prefix`

| AZURE
| `az://vault-uri/prefix`. The `vault-uri` value should not contain a `https://` prefix.

| GCP
| `gcp://projectID/prefix`

| Redis
| `redis://secret:redis@example.com`

|===

> == Set secrets using shell commands

Author: @Jeffail. Is this heading accurate?

Reviewer: Yeah this works
